### ISSUE TITLE
Integrity constraint validation with project config

### ISSUE THREAD

**Issue by wallacio** [mmikkel/Reasons-Craft3#9]
I've spent a while trying to debug why a deployment has been failing. My site uses project-config.

Craft: 3.5.11.1
Reasons: 2.2.1
DB Engine: MySQL

Steps to reproduce:
- git pull of my project, which contains project config
- composer install
- craft setup

The final part of craft setup is applying the project config, but this fails with the SQL error below. Removing the `reasons_conditionals` folder from config/project allows the installation to complete successfully.

Comparing the databases, the value for `fieldLayoutId` (0) that's being inserted here relates to a different/incompatible field layout on this installation. My guess is that fieldLayoutId shouldn't be in the table at all, and the UUID used instead? There's apparently no fieldLayoutId in the yaml.

Aside from this, I can't get Reasons to work on this version of Craft even when it is configured correctly!

``
Applying changes from your project config files ... error: SQLSTATE[23000]: Integrity constraint violation: 1452 Cannot add or update a child row: a foreign key constraint fails (`craft3`.`craft_reasons`, CONSTRAINT `craft_reasons_fieldLayoutId_fk` FOREIGN KEY (`fieldLayoutId`) REFERENCES `craft_fieldlayouts` (`id`) ON DELETE CASCADE ON UPDATE CASCADE)
The SQL being executed was: INSERT INTO `craft_reasons` (`fieldLayoutId`, `conditionals`, `uid`, `dateCreated`, `dateUpdated`) VALUES (0, '{\"2e8dd7f7-f84f-4c0a-b3b8-8ac1bce79d9d\":[[{\"field\":\"e99e7f78-3ce1-4e44-b165-a8edbc360aca\",\"compare\":\"==\",\"value\":\"true\"}]],\"9fcbd106-b808-47ec-942b-668235124254\":[[{\"field\":\"e99e7f78-3ce1-4e44-b165-a8edbc360aca\",\"compare\":\"==\",\"value\":\"true\"}]],\"6d8fcd42-634a-4dac-b682-ede47aba83f0\":[[{\"field\":\"e99e7f78-3ce1-4e44-b165-a8edbc360aca\",\"compare\":\"==\",\"value\":\"true\"}]],\"c01c6210-ed29-4fe6-a390-87eef631095c\":[[{\"field\":\"e99e7f78-3ce1-4e44-b165-a8edbc360aca\",\"compare\":\"==\",\"value\":\"true\"}]],\"19ced37f-881d-491f-898b-91ca9de7057e\":[[{\"field\":\"e99e7f78-3ce1-4e44-b165-a8edbc360aca\",\"compare\":\"==\",\"value\":\"true\"}]],\"063f327a-3e45-4cdf-b47c-43359485a7bc\":[[{\"field\":\"e99e7f78-3ce1-4e44-b165-a8edbc360aca\",\"compare\":\"==\",\"value\":\"true\"}]]}', '2ca19fa0-117f-41d6-81f2-3f2fb1e1c6dd', '2020-09-21 21:25:33', '2020-09-21 21:25:33')
``

**Comment by mmikkel** [comment]
Hi @wallacio, finally had the time to look at this. Thanks again for sending me your files, which helped me track down this issue.

I uncovered a bug in Reasons where, if installing Craft using existing Yaml files, Project Config could attempt to save conditionals to the database before saving the conditionals' related field layouts – which would result in an integrity constraint violation when applying the Yaml files. **This issue has been resolved in [Reasons 2.2.2](https://github.com/mmikkel/Reasons-Craft3/releases/tag/2.2.2).**

Note that in your case, applying the Yaml files is still going to fail w/ Reasons 2.2.2, because your `reasons_conditionals` Yaml files refer to field layouts which don't actually exist in your Project Config. In other words, your project config files are basically corrupted. This could be due to a different bug in Reasons, where deleting field layouts (e.g. if an entry type was deleted) would not actually delete any related conditionals from the Project Config. This has also been fixed in Reasons 2.2.2.

To fix your project config files (i.e. by removing the `reasons_conditionals` files that don't refer to actual field layouts) you can either

a) Run `clear-caches/all` and then `project-config/rebuild` in the original project, to regenerate the Project Config yaml files. Since Reasons 2.2.2, any conditionals that don't refer to field layouts currently in the `fieldlayouts` database table will be removed from the Project Config (both from the `projectconfig` database table and from the `config/project` Yaml files.

b) Simply delete the offending Yaml files from your `config/project` folder.

**Comment by slavanossar** [comment]
Experiencing something similar when trying to add Reasons to an existing site. Package installs correctly, but then all pages display the error (presumably due to config rebuild as `project-config/rebuild` also throws same error):

```error: SQLSTATE[42703]: Undefined column: 7 ERROR:  column reasons.fieldlayoutid does not exist
LINE 3: ...fieldlayouts" "fieldlayouts" ON fieldlayouts.id = reasons.fi...
                                                             ^
HINT:  Perhaps you meant to reference the column "reasons.fieldLayoutId".
The SQL being executed was: SELECT "reasons"."uid", "reasons"."conditionals", "fieldlayouts"."uid" AS "fieldLayoutUid"
FROM "reasons" "reasons"
INNER JOIN "fieldlayouts" "fieldlayouts" ON fieldlayouts.id = reasons.fieldLayoutId```

Can send through a DB dump and the `project/config` directory if needed 👍

Craft: Pro 3.5.14
Reasons: 2.2.2
DB Engine: PostgreSQL 12.1
